router/router.go

```diff
@@ -80,6 +80,7 @@ type routeSet struct {
 	// The following state is computed from 'routes'.
 	prefixHash    map[string]*route // Keys are everything up to the first open parenthesis character '('
 	prefixLengths []int             // Descending list of unique prefix lengths
+	hostHash      map[string]bool
```

Comment:
I don't think it makes sense to only store 'useProxy' here. Rather make the map from hostname -> *target. The variable in this case deserves a better name, such as 'hostToTarget'.